File: scripts/patchcheck.cpp
/*
 * Copyright (C) 2004-2026 Metaphonic Labs
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General
 * Public License along with this program; if not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/*
 * patchcheck -- what a .patch means, held still.
 *
 * src/PatchFile.h is the format with no file and no synth around it, which is
 * what makes this possible at all: the rules used to live inside a loop over
 * fgets() that needed a real file on a real disk and a real channel to load
 * onto, so the only thing that ever exercised them was the application
 * starting up. They are an ordinary function over an ordinary string now, and
 * this runs everywhere gencheck does.
 *
 * Three things are checked.
 *
 *   The corpus. Every shipped .patch parses, names a graph, and says nothing
 *   the reader has to complain about. That is the claim "no corpus file
 *   changed meaning" -- a .patch that started complaining is a .patch this
 *   reader disagrees with the old one about.
 *
 *   The round trip. parse -> compose -> parse, and the second document equals
 *   the first, for every shipped file and every fixture. That is what lets a
 *   browser write a .patch the desktop reads back, and what stops compose
 *   from being judged by eye.
 *
 *   The awkward lines. A fixture per documented answer, asserting the answer
 *   rather than whatever the code happens to do -- an `info' with no value, a
 *   line with no space in it, a line that starts with a NUL, CRLF endings,
 *   `side 99', `side' naming its own channel, an `fx.' name nothing declares,
 *   and `name 1,2,3'. Every one of those is a case the two old parsers
 *   answered differently, and several are cases the corpus cannot contain.
 *
 * What is not checked here is whether the graph exists, whether anything
 * declares `cutoff', or whether 99 is a channel. Those need a synth and are
 * PatchApply's; scripts/dspcheck --patch is what loads the corpus onto one.
 *
 *     scripts/patchcheck [corpus-dir]
 *
 * With no argument it checks the fixtures alone, which is how CI runs it on
 * the platforms that have no source tree beside the build. Exit status is the
 * number of failures.
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <algorithm>
#include <filesystem>
#include <fstream>
#include <sstream>

#include "PatchFile.h"

using std::ifstream;
using std::ostringstream;

static int failed = 0;

static void ok (const string &what)
{
    printf("ok    %s\n", what.c_str());
}

static void fail (const string &what, const string &detail)
{
    printf("FAIL  %s%s%s\n", what.c_str(), detail.empty() ? "" : ": ",
           detail.c_str());
    failed++;
}

static void check (bool good, const string &what, const string &detail = "")
{
    if (good)
        ok(what);
    else
        fail(what, detail);
}

static void checkEq (const string &got, const string &want, const string &what)
{
    check(got == want, what, "got `" + got + "', wanted `" + want + "'");
}

/* ------------------------------------------------------------------ */
/* Two documents are the same document                                  */
/* ------------------------------------------------------------------ */

/* Everything a file can say, compared field by field. Composing both and
 * comparing the text would be shorter and would also be circular -- it would
 * pass for any two documents compose happened to flatten together, which is
 * precisely the failure the round trip is looking for.
 *
 * complaints are not compared. A document that complained once must compose
 * to a file that does not, so the second pass having none is the property,
 * and it is asserted separately below.
 */
static bool sameDoc (const thPatchDoc &a, const thPatchDoc &b, string &why)
{
    if (a.dsp != b.dsp)
    {
        why = "dsp `" + a.dsp + "' became `" + b.dsp + "'";
        return false;
    }

    if (a.effect != b.effect)
    {
        why = "effect `" + a.effect + "' became `" + b.effect + "'";
        return false;
    }

    if (a.side != b.side)
    {
        char buf[64];

        snprintf(buf, sizeof(buf), "side %d became %d", a.side, b.side);
        why = buf;
        return false;
    }

    /* An info property with an empty value is not written, so it is not read
       back either -- parse refuses to make one, which is what makes this a
       plain comparison rather than one with an exception in it. */
    if (a.info != b.info)
    {
        why = "the info properties differ";
        return false;
    }

    if (a.args.size() != b.args.size())
    {
        why = "a different number of args came back";
        return false;
    }

    for (map<string, vector<float> >::const_iterator j = a.args.begin();
         j != a.args.end(); ++j)
    {
        map<string, vector<float> >::const_iterator k = b.args.find(j->first);

        if (k == b.args.end())
        {
            why = "`" + j->first + "' was lost";
            return false;
        }

        if (j->second.size() != k->second.size())
        {
            why = "`" + j->first + "' changed how many values it has";
            return false;
        }

        for (size_t i = 0; i < j->second.size(); i++)
            if (j->second[i] != k->second[i])
            {
                char buf[128];

                snprintf(buf, sizeof(buf), "%s[%u]: %f became %f",
                         j->first.c_str(), (unsigned)i, j->second[i],
                         k->second[i]);
                why = buf;
                return false;
            }
    }

    return true;
}

/* parse, compose, parse: the same document, and a file that has stopped
   complaining. Returns false having already said what went wrong. */
static bool roundTrips (const string &text, const string &what)
{
    thPatchDoc first, second;
    string why;

    if (!thPatchParse(text, first, why))
    {
        fail(what + ": parses", why);
        return false;
    }

    const string written = thPatchCompose(first);

    if (!thPatchParse(written, second, why))
    {
        fail(what + ": what it writes parses", why);
        return false;
    }

    if (!sameDoc(first, second, why))
    {
        fail(what + ": round trips", why);
        return false;
    }

    /* Whatever the reader had to skip on the way in, it wrote nothing that
       has to be skipped on the way out. A compose that preserved a bad line
       would round-trip and still be wrong. */
    if (!second.complaints.empty())
    {
        fail(what + ": what it writes is clean", second.complaints[0]);
        return false;
    }

    /* And again, so that "the bytes are stable" is a claim about bytes and
       not only about what they mean. This is the property a page needs to
       offer a download that a desktop will not rewrite on its first save. */
    if (thPatchCompose(second) != written)
    {
        fail(what + ": writes the same bytes twice", "");
        return false;
    }

    ok(what + ": round trips");

    return true;
}

/* ------------------------------------------------------------------ */
/* The corpus                                                           */
/* ------------------------------------------------------------------ */

static bool readFile (const std::filesystem::path &p, string &out)
{
    ifstream in(p, std::ios::binary);

    if (!in)
        return false;

    ostringstream buf;

    buf << in.rdbuf();
    out = buf.str();

    return true;
}

static void checkCorpus (const string &dir)
{
    std::error_code ec;
    int seen = 0;

    std::filesystem::recursive_directory_iterator walk(dir, ec);

    if (ec)
    {
        fail("the corpus is there", dir + ": " + ec.message());
        return;
    }

    /* Sorted, so a failure names the same file on every machine: the
       directory iterator's order is the filesystem's. */
    vector<std::filesystem::path> files;

    for (const std::filesystem::directory_entry &e : walk)
        if (e.is_regular_file() && e.path().extension() == ".patch")
            files.push_back(e.path());

    std::sort(files.begin(), files.end());

    for (const std::filesystem::path &p : files)
    {
        const string name = p.filename().string();
        string text;

        seen++;

        if (!readFile(p, text))
        {
            fail(name + ": reads", "could not be opened");
            continue;
        }

        thPatchDoc doc;
        string why;

        if (!thPatchParse(text, doc, why))
        {
            fail(name + ": parses", why);
            continue;
        }

        check(!doc.dsp.empty(), name + ": names a graph");

        /* The one that matters for "no shipped file changed meaning": this
           reader understood every line of it. */
        check(doc.complaints.empty(), name + ": has nothing to complain about",
              doc.complaints.empty() ? "" : doc.complaints[0]);

        roundTrips(text, name);
    }

    check(seen > 0, "the corpus has files in it");
}

/* ------------------------------------------------------------------ */
/* The awkward lines                                                    */
/* ------------------------------------------------------------------ */

/* Parses, and the caller gets the document. Fails the named check and
   returns false otherwise, so each fixture below reads as its assertion. */
static bool parses (const string &text, thPatchDoc &doc, const string &what)
{
    string why;

    if (thPatchParse(text, doc, why))
        return true;

    fail(what, why);

    return false;
}

static void checkFixtures (void)
{
    /* A patch is its `dsp' line. Nothing else is required and nothing else
       can fail it. */
    {
        thPatchDoc doc;
        string why;

        check(!thPatchParse("cutoff 1.0\n", doc, why),
              "a file with no dsp line is not a patch");
        check(why == "names no dsp", "and says so", why);

        check(!thPatchParse("", doc, why), "nor is an empty one");

        if (parses(" \t dsp ts1.dsp\n", doc, "a dsp line alone is a patch"))
        {
            ok("a dsp line alone is a patch");
            checkEq(doc.dsp, "ts1.dsp", "leading whitespace is not the name");
        }
    }

    /* `info foo' -- a property named and given nothing. The desktop refused
       the whole file over this, sharing its exit with a DSP that would not
       load, so a typo in a comment cost the instrument. A complaint, and the
       rest of the file is read. */
    {
        thPatchDoc doc;

        if (parses("dsp ts1.dsp\ninfo foo\ncutoff 2.5\n", doc,
                   "an info line with no value does not fail the patch"))
        {
            ok("an info line with no value does not fail the patch");
            check(doc.info.empty(), "and invents no property");
            check(doc.args.count("cutoff") == 1,
                  "and the line after it is still read");
            check(doc.complaints.size() == 1, "and it is complained about");
        }
    }

    /* A word on a line by itself. Both old parsers skipped it in silence;
       silence is what made a misspelling impossible to find. */
    {
        thPatchDoc doc;

        if (parses("dsp ts1.dsp\ncutoff\n", doc,
                   "a line with no space in it is not an arg"))
        {
            ok("a line with no space in it is not an arg");
            check(doc.args.empty(), "and sets nothing");
            check(doc.complaints.size() == 1, "and is complained about");
        }
    }

    /* A line whose first byte is a NUL. strlen() said 0, so the desktop's
       `buffer[len - 1]' was `buffer[-1]' -- a write one byte in front of the
       buffer, which is why the length is checked there to this day. A string
       has a length that does not end at a NUL, so this is now only a line
       with a strange character at the front of it. */
    {
        const string text = string("dsp ts1.dsp\n") + string(1, '\0') +
                            "cutoff 2.5\n";
        thPatchDoc doc;

        if (parses(text, doc, "a line starting with a NUL is survived"))
        {
            ok("a line starting with a NUL is survived");
            checkEq(doc.dsp, "ts1.dsp", "and the patch is still a patch");
        }
    }

    /* CRLF. The desktop kept the CR, so `dsp ts1.dsp\r' named a file with a
       control character in it -- a load that failed reporting a name that
       looked exactly right. */
    {
        thPatchDoc doc;

        if (parses("dsp ts1.dsp\r\ninfo title Super Res\r\ncutoff 2.5\r\n",
                   doc, "a file with CRLF endings parses"))
        {
            ok("a file with CRLF endings parses");
            checkEq(doc.dsp, "ts1.dsp", "with no CR in the graph's name");
            checkEq(doc.info["title"], "Super Res",
                    "nor at the end of a property");
            check(doc.args["cutoff"].size() == 1 &&
                  doc.args["cutoff"][0] == 2.5f, "nor after a value");
        }
    }

    /* `side' is 1-based in the file and engine numbering in the document, and
     * that is the whole of what happens to it here. Whether 99 is a channel,
     * and whether a channel may name itself, need to know how many channels
     * there are and which one this is -- so they are PatchApply's, and
     * keeping the number is what makes the round trip exact. */
    {
        thPatchDoc doc;

        if (parses("dsp ts1.dsp\nside 99\neffect fx/delay.dsp\n", doc,
                   "side is read 1-based"))
        {
            ok("side is read 1-based");
            check(doc.side == 98, "and kept even where no such channel is");
        }

        if (parses("dsp ts1.dsp\nside 4\neffect fx/delay.dsp\n", doc,
                   "a side names a channel"))
            check(doc.side == 3, "a side is one less than what is written");

        if (parses("dsp ts1.dsp\nside 0\neffect fx/delay.dsp\n", doc,
                   "side 0 is no side"))
            check(doc.side == -1, "side 0 is no side");

        if (parses("dsp ts1.dsp\nside -1\neffect fx/delay.dsp\n", doc,
                   "a negative side is no side"))
            check(doc.side == -1, "a negative side is no side");
    }

    /* The effect line itself -- the divergence that cost the page every
       channel effect. Its value is a filename, so a reader that tries it as a
       number gets NaN and drops the line, which is exactly what patch.js
       did. */
    {
        thPatchDoc doc;

        if (parses("dsp ts1.dsp\neffect fx/delay.dsp\nfx.wet 0.5\n", doc,
                   "an effect line is read"))
        {
            ok("an effect line is read");
            checkEq(doc.effect, "fx/delay.dsp", "by the name it was given");
            check(doc.args.count("fx.wet") == 1,
                  "and its parameters are args like any other");
            check(doc.args.count("effect") == 0,
                  "and the effect itself is not an arg");
            check(doc.args.count("dsp") == 0, "nor is the graph");
            check(doc.args.count("side") == 0, "nor is the side");
        }
    }

    /* An `fx.' name nothing declares is a document like any other: the
     * document does not know what an effect declares, and refusing one here
     * would mean refusing every `fx.' line in a file read without a synth.
     * Where it is refused is PatchApply, by name and with a message.
     *
     * What this does assert is that the values follow their effect out of
     * compose: an `fx.' line with no `effect' above it is a line the reader
     * has nowhere to put, so writing one would be writing a patch that
     * complains at itself on every load.
     */
    {
        thPatchDoc doc;

        if (parses("dsp ts1.dsp\nfx.nosuchthing 0.5\n", doc,
                   "an fx. name with no effect parses"))
        {
            ok("an fx. name with no effect parses");
            check(doc.args.count("fx.nosuchthing") == 1,
                  "and is kept as an arg");
            check(thPatchCompose(doc).find("fx.nosuchthing") == string::npos,
                  "and is not written back without an effect to hold it");
        }
    }

    /* `name 1,2,3'. The one divergence resolved toward the page: a thArg has
       always held a list and the desktop kept the first value only. */
    {
        thPatchDoc doc;

        if (parses("dsp ts1.dsp\nwave 1,2,3\n", doc,
                   "a multi-value arg keeps every value"))
        {
            ok("a multi-value arg keeps every value");
            check(doc.args["wave"].size() == 3, "all three of them");

            if (doc.args["wave"].size() == 3)
                check(doc.args["wave"][0] == 1 && doc.args["wave"][1] == 2 &&
                      doc.args["wave"][2] == 3, "in the order written");

            check(thPatchCompose(doc).find("wave 1.000000,2.000000,3.000000") !=
                  string::npos, "and writes them back as a list");
        }
    }

    /* A value that is not a number. strtof took what it could and shrugged at
       the rest, so `cutoff abc' was a silent zero -- a parameter set to a
       number nobody typed, indistinguishable from one somebody meant. */
    {
        thPatchDoc doc;

        if (parses("dsp ts1.dsp\ncutoff abc\nres 4abc\nq 1,,3\n", doc,
                   "a value that is not a number is not a zero"))
        {
            ok("a value that is not a number is not a zero");
            check(doc.args.empty(), "and sets nothing");
            check(doc.complaints.size() == 3, "and each line is complained about");
        }
    }

    /* Comments, blank lines and the escape. */
    {
        thPatchDoc doc;

        if (parses("# a comment\n\n   # an indented one\ndsp ts1.dsp\n"
                   "info comments one\\ntwo\n", doc,
                   "comments and blank lines are skipped"))
        {
            ok("comments and blank lines are skipped");
            check(doc.complaints.empty(), "in silence",
                  doc.complaints.empty() ? "" : doc.complaints[0]);
            checkEq(doc.info["comments"], "one\ntwo",
                    "and an escaped newline is a newline");
        }
    }

    /* A file with no trailing newline: the last line is still a line. */
    {
        thPatchDoc doc;

        if (parses("dsp ts1.dsp\ncutoff 2.5", doc,
                   "a file with no trailing newline is whole"))
        {
            ok("a file with no trailing newline is whole");
            check(doc.args.count("cutoff") == 1, "and its last line counts");
        }
    }

    /* And every fixture that is a patch round-trips, which is the property
       the corpus cannot test: no shipped file has an effect, a side, a
       multi-value arg or an escaped newline in it. */
    roundTrips("dsp ts1.dsp\nside 4\neffect fx/delay.dsp\n"
               "info title Super Res\ninfo comments one\\ntwo\n"
               "cutoff 2.5\nwave 1,2,3\nfx.wet 0.5\n",
               "a patch using every line the format has");
}

int main (int argc, char *argv[])
{
    checkFixtures();

    if (argc > 1)
        checkCorpus(argv[1]);
    else
        printf("(no corpus directory given; fixtures only)\n");

    printf("\n%d failure%s\n", failed, failed == 1 ? "" : "s");

    return failed;
}

File: wasm/web/patchcheck.mjs
/*
 * Copyright (C) 2004-2026 Metaphonic Labs
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General
 * Public License along with this program; if not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/*
 * patchcheck.mjs -- a .patch, read by the browser module, held against the
 * native one.
 *
 *   node wasm/web/patchcheck.mjs [BUILD_DIR] [NATIVE_BUILD_DIR]
 *
 * The format is read once (src/PatchFile.h) and compiled twice, and the
 * point of reading it once is that the two readings are of the same thing.
 * Nothing enforces that by itself, and the last time nothing enforced it the
 * two readings drifted: the page dropped every `effect' line on the floor and
 * sent `side' to the engine as a chanarg. It is the divergence the whole
 * extraction exists to stop, so it is a gate.
 *
 * So scripts/patchcheck -j writes each awkward fixture and the document it
 * read from it; this hands the same bytes to tw_patch_read and compares the
 * two dumps byte for byte. The refusals are compared too -- text that is not
 * a patch has to be not a patch on both sides, and for the same stated
 * reason.
 *
 * Then the half a dump cannot show: that tw_patch_apply actually puts one on
 * a channel -- the graph it names, and its overrides on that graph's
 * chanargs -- and that a patch naming a graph the module does not have is
 * refused with the channel left alone.
 *
 * Exit status is the number of failures.
 */

import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { execFileSync } from 'node:child_process';
import { fileURLToPath, pathToFileURL } from 'node:url';

const here = path.dirname(fileURLToPath(import.meta.url));
const top = path.join(here, '..', '..');
const build = path.resolve(process.argv[2] ?? path.join(top, 'build-web'));
const native = path.resolve(process.argv[3] ?? path.join(top, 'build'));

const harness = path.join(native, 'scripts', 'patchcheck');

let failures = 0;

const ok = (what) => process.stdout.write(`ok    ${what}\n`);

const fail = (what, detail = '') =>
{
    process.stdout.write(`FAIL  ${what}${detail ? `: ${detail}` : ''}\n`);
    failures++;
};

const check = (good, what, detail = '') =>
    good ? ok(what) : fail(what, detail);

if (!fs.existsSync(harness))
{
    process.stdout.write(
        `patchcheck: no patchcheck in ${native}; build the native tree ` +
        'first -- it is what the module is compared against.\n');
    process.exit(1);
}

/* The native side first: it writes each fixture's bytes as well as its
   document, and the module reads that exact text. */
const scratch = fs.mkdtempSync(path.join(os.tmpdir(), 'patchcheck-'));

execFileSync(harness, ['-j', scratch],
             { stdio: ['ignore', 'ignore', 'inherit'] });

const names = fs.readFileSync(path.join(scratch, 'index.txt'), 'utf8')
    .split('\n').filter((n) => n !== '');

check(names.length > 0, 'the native harness wrote its fixtures');

const { default: createThinkWeb } =
    await import(pathToFileURL(path.join(build, 'thinkweb.js')).href);

const log = [];
const M = await createThinkWeb({
    print: (s) => log.push(s),
    printErr: (s) => log.push(s),
});

/* The same rate and window the page's own synth is made with (host.js).
   Nothing a document says depends on either, which is the point: if a
   reading ever starts to, this is where it shows. */
M._tw_create(44100, 256, 4096);

/* ---- the reading ---- */

for (const name of names)
{
    const text = fs.readFileSync(path.join(scratch, `${name}.patch`), 'utf8');

    /* The native dump is the document and then the refusal, a line each --
       one of the two is always empty. */
    const [wantJson, wantWhy] =
        fs.readFileSync(path.join(scratch, `${name}.json`), 'utf8')
            .split('\n');

    const gotJson = M.ccall('tw_patch_read', 'string', ['string'], [text]);
    const gotWhy = M.ccall('tw_patch_why', 'string', [], []);

    check(gotJson === wantJson, `${name}: reads the same`,
          gotJson === wantJson ? ''
              : `\n    native ${wantJson}\n    wasm   ${gotJson}`);
    check(gotWhy === wantWhy, `${name}: refuses for the same reason`,
          `native "${wantWhy}", wasm "${gotWhy}"`);
}

/* ---- the first-run configuration ---- */

/* The four patches a channel gets when nothing has aimed it, and the rule
 * for a channel above the last of them. The page had a copy of both, with a
 * comment saying the list was gthPrefs.cpp's "exactly" -- which is the kind
 * of promise nothing keeps, so now it is this.
 */
{
    const wanted = fs.readFileSync(path.join(scratch, 'defaults.txt'), 'utf8')
        .split('\n').slice(0, -1);

    const got = wanted.map((_, c) =>
        M.ccall('tw_patch_default', 'string', ['number'], [c]));

    check(got.join('\n') === wanted.join('\n'),
          'the first-run defaults are the same on both sides',
          `\n    native ${wanted.join(', ')}\n    wasm   ${got.join(', ')}`);

    check(M._tw_patch_default_count() === new Set(wanted).size,
          'and there are as many of them as there are distinct names');
}

/* ---- and onto a channel ---- */

/* The graph a shipped patch names, and the shipped patch itself. The module
   looks a `dsp' line up among the instruments the page handed over, which is
   the same lookup the application makes under DSP_PATH -- so handing this one
   graph over stands in for the page's Start.

   The corpus rather than a fixture, because what is being asked here is
   whether a real .patch reaches a real channel: the fixtures above have
   already said everything there is to say about awkward text. */
M.ccall('tw_instrument', 'number', ['string', 'string'],
        ['ts1.dsp',
         fs.readFileSync(path.join(top, 'dsp', 'ts1.dsp'), 'utf8')]);

const SHIPPED = fs.readFileSync(
    path.join(top, 'patches', 'leads', 'SuperRes.patch'), 'utf8');

const apply = (channel, text, name = '') =>
    M.ccall('tw_patch_apply', 'number', ['number', 'string', 'string'],
            [channel, text, name]) !== 0;

check(apply(0, SHIPPED, 'leads/SuperRes.patch'),
      'a shipped patch goes on a channel');

check(M.ccall('tw_patch_why', 'string', [], []) === '',
      'and has nothing to say about it');

/* The overrides landed, which is the half of loading a patch that used to be
   a separate call the page made in the right order by hand. */
const json = M.ccall('tw_patch_json', 'string', ['number'], [0]);

check(json !== '', 'and the channel says what is on it');

if (json !== '')
{
    const doc = JSON.parse(json);

    check(doc.dsp === 'ts1.dsp', 'by the name the file gave',
          `got ${doc.dsp}`);
    /* 1.040810, the `cutoff' line in SuperRes.patch, through a float and
       back out as a double: the value the file gives, not a rounding of it
       and not the 4 the graph declares. */
    check(doc.args.cutoff?.[0] === Math.fround(1.040810),
          'with its overrides in it',
          `got ${JSON.stringify(doc.args.cutoff)}`);
}

/* ---- and what the slot remembers ---- */

/* The three things a slot knows that the file does not. They are what the
 * page draws its channel row from, and what the desktop has been able to say
 * for twenty years: which file this is, whether it has been edited since it
 * was read, and which load it is.
 */
if (json !== '')
{
    const slot = JSON.parse(json);

    check(slot.name === 'leads/SuperRes.patch',
          'the slot is named by the file it came from', `got ${slot.name}`);
    check(slot.dirty === false, 'and a patch just read is not edited');
    check(slot.generation > 0, 'and carries a generation',
          `got ${slot.generation}`);

    /* A chanarg edit is what makes it dirty, and it arrives the way every
       edit on the page does: as a command, applied through tw_panel_edit. */
    const edited = M.ccall('tw_panel_edit', 'number',
                           ['number', 'number', 'number', 'string', 'string'],
                           [0, 0, 0, 'cutoff', '2.0']);

    check(edited === 1, 'moving a control is an edit',
          M.ccall('tw_panel_why', 'string', [], []));
    check(M.ccall('tw_patch_dirty', 'number', ['number'], [0]) === 1,
          'and the channel says its patch has been edited');

    const after = JSON.parse(
        M.ccall('tw_patch_json', 'string', ['number'], [0]));

    check(after.dirty === true, 'which is in the row the page draws');
    check(after.generation === slot.generation,
          'and editing one is not loading another');
}

/* ---- and back out ---- */

/* The bytes a Save would write, which is the first time a browser can save a
 * patch at all. They are the bytes the desktop writes -- both halves are
 * shared -- so what is asserted here is that they come back in: compose,
 * read, and the document is the one the channel holds, edit and all.
 */
{
    const text = M.ccall('tw_patch_compose', 'string', ['number', 'string'],
                         [0, 'Wed Jan  1 00:00:00 2025']);

    check(text.startsWith('# Thinksynth Patch File\n'),
          'a channel composes back into a .patch');

    const back = M.ccall('tw_patch_read', 'string', ['string'], [text]);

    check(back !== '', 'which reads as a patch');

    if (back !== '')
    {
        const doc = JSON.parse(back);
        const now = JSON.parse(
            M.ccall('tw_patch_json', 'string', ['number'], [0]));

        check(doc.dsp === now.dsp, 'naming the graph the channel is playing',
              `got ${doc.dsp}`);

        /* The edit above, not the 1.040810 the file gave: what a Save is for
           is writing down what somebody has moved. */
        check(doc.args.cutoff?.[0] === 2, 'with the value that was typed, ' +
              'not the one the file gave', `got ${doc.args.cutoff}`);
    }

    /* And a page that kept them says so, which is what puts the mark out. */
    check(M.ccall('tw_patch_saved', 'number', ['number', 'string'],
                  [0, 'mine.patch']) === 1, 'a patch can be marked saved');
    check(M.ccall('tw_patch_dirty', 'number', ['number'], [0]) === 0,
          'and is no longer edited');
    check(JSON.parse(M.ccall('tw_patch_json', 'string', ['number'], [0]))
              .name === 'mine.patch',
          'under the name it was saved as');

    check(M.ccall('tw_patch_compose', 'string', ['number', 'string'],
                  [9, '']) === '',
          'a channel with nothing on it composes nothing');
}

/* A graph nothing has handed over. The refusal is the point, and so is what
   it leaves behind: the channel is not touched, so whatever was playing is
   still playing and the patch that is recorded there is still the last one
   that loaded. */
check(!apply(0, 'dsp nosuchgraph.dsp\n'),
      'a patch naming a graph the module does not have is refused');

check(M.ccall('tw_patch_why', 'string', [], []) !== '', 'and says why');

const untouched = M.ccall('tw_patch_json', 'string', ['number'], [0]);

check(untouched === M.ccall('tw_patch_json', 'string', ['number'], [0]) &&
      JSON.parse(untouched).generation === JSON.parse(json).generation,
      'and leaves the channel as it was');

/* Text that is not a patch at all reaches here as a refusal and not as a
   half-loaded channel. */
check(!apply(1, 'this is not a patch\n'),
      'text that names no graph is refused');

check(M.ccall('tw_patch_why', 'string', [], []) === 'names no dsp',
      'with the reading\'s own words');

process.stdout.write(`\n${failures} failure${failures === 1 ? '' : 's'}\n`);

if (failures > 0 && log.length > 0)
    process.stdout.write(`\nthe module said:\n${log.join('\n')}\n`);

process.exit(failures);
